#!/bin/bash
# Script to subtract bed file from output table generated by Tandem Repeat Finder
# Expects TRF output file (.dat) and a txt file with a list of the sequences (fasta headers) 

###########################################
TRFFILE=myautosomes.fa.2.7.7.80.10.50.6.dat
MYSCAFFOLDS=myscaffolds.txt
###########################################


if [ -f allmt.allscaf.bed ];
 then
 rm allmt.allscaf.bed
 fi
touch allmt.allscaf.bed

grep -n 'Sequence' $TRFFILE | cut -f1 -d ':' > mylinenumbers.txt
wc -l $TRFFILE | cut -f1 -d ' ' >> mylinenumbers.txt

nscaffolds=$(wc -l $MYSCAFFOLDS | cut -f1 -d ' ') 
 
for myindex in $(seq 1 ${nscaffolds})
 do
 myindex2=$(( $myindex + 1 ))
 myscaffold=$(awk -v myline="${myindex}" 'NR == myline' ${MYSCAFFOLDS})
 myline1=$(awk -v myline="${myindex}" 'NR == myline' mylinenumbers.txt)
 myline2=$(awk -v myline="${myindex2}" 'NR == myline' mylinenumbers.txt) 
 echo $myscaffold
 awk -v mystart="${myline1}" -v myend="${myline2}"  'NR > mystart && NR < myend' OFS="\t" $TRFFILE > myscaffold.dat
 # define here period length (column 3) and repeat range (column 4) and repeat score (column 6):
 # 19 or 20 dinucleotide repeats:
 #sed 's/ /\t/g' myscaffold.dat | awk '$3==2' | awk '$4==19 || $5==20' OFS="\t" | awk '$6==100' OFS="\t" > myscaffold.subset.dat 
 # 7-9 tetranucleotide repeats:
 sed 's/ /\t/g' myscaffold.dat | awk '$3==4' | awk '$4==7 || $4==8 || $4==9' OFS="\t" | awk '$6==100' OFS="\t" > myscaffold.subset.dat
 cut -f1,2,14 myscaffold.subset.dat > allmt.bed
 sed -i "s/^/${myscaffold}\t/" allmt.bed
 awk '$4 = $1"_"$2' OFS="\t" allmt.bed > allmt2.bed
 awk '$5 = $3 - $2' OFS="\t" allmt2.bed > allmt3.bed 
 cut -f3,14 myscaffold.subset.dat > allmt4.bed
 paste allmt3.bed allmt4.bed > allmt5.bed
 awk '$8 = $7' OFS="\t" allmt5.bed > allmt6.bed
 awk '$9 = $7' OFS="\t" allmt6.bed > allmt7.bed
 cat allmt7.bed >> allmt.allscaf.bed
 rm allmt.bed allmt2.bed allmt3.bed allmt4.bed allmt5.bed allmt6.bed allmt7.bed myscaffold.dat myscaffold.subset.dat
done


# Optionally, include in the bed file flanking regions of a certain size:
echo "Flanking regions..."
awk '$2 = $2-20' OFS="\t" allmt.allscaf.bed | awk '$3 = $3+20' OFS="\t" > allmt.allscaf.flanking.bed 

echo "Total number of selected short tandem repeats:"
wc -l allmt.allscaf.bed | cut -f1 -d ' '

echo "Total combined length of selected short tandem repeats:"
awk '{SUM+=$5}END{print SUM}' allmt.allscaf.bed

echo "Data stored in file 'allmt.allscaf.bed'."



# Note, although indeed there are many short tandem repeats in the genome, their actual percentage of the entire genome is relatively small.
# For our example, for the brown bear genome I found a total combined length of 10Mb.
# This is less than 0.5% of the genome.
