# MicrosatelliteGenotypingFromBam

With the advent of next generation sequencing, SNPs (single nucleotide polymorphisms) have replaced microsatellites as the mainstream genetic marker. These SNP data sets are typically generated by mapping short sequencing reads to a reference genome (which generates bam-files), by subsequently using the information in the bam-files to call genotypes (i.e., determine for each site whether it is most likely homozygous reference, heterozygous, or homozygous alternative), and by lastly selecting the sites where for the given subset of individuals two or more alleles (i.e., A,C,G,T) are present. 

However, instead of extracting SNP data from the information stored in bam-files, why not extract microsatellite data? Given the relatively high mutation rate and allelic richness of microsatellites, one could even argue that microsatellites are a more informative genetic marker than biallelic SNPs. While traditional microsatellite studies are based on a limited number of markers (typically less than 30), whole genome resequencing data potentially contains information for many thousands of microsatellites.   

## Workflow overview

The pipeline presented here extracts microsatellite data from input bam-files. The prerequisite is that the short read data has been mapped against a reference genome which has not been masked for repetitive genomes. Furthermore, the sequencing data should at least be 10x (preferably 15x or more), and the read lengths preferably not shorter than 150 bp.

The procedure to produce a microsatellite data set from bam-files involves computer work only (not wet lab stuff), and consists of the following five steps:

1. microsatellite loci selection: detect and select a set of microsatellite loci in a reference genome
2. microsatellite data extraction: extract the selected microsatellite loci from all individual genomes
3. allele scoring: count for each microsatellite locus the number of reads supporting each allele  
4. genotype calling: convert count data into genotypes using a decision tree
5. data quality filtering: filter individuals and loci on levels of missing data  

Albeit entirely computer-based, in a way each step of this pipeline resembles steps in the traditional wet-lab protocol of a microsatellite study, namely: 1.) primer design, 2.) target-specific DNA-amplification through PCR, 3.) allele scoring through capillary electrophoresis, 4.) genotype calling using a decision tree (e.g., distinguishing true signals from stutter alleles and null alleles), and .5) data quality filtering.     

## Step 1. Select a set of microsatellite loci

Microsatellites in a reference genome can be detected using the software TRFfinder, by running the command:

./trf409.linux64 mygenome.fa -h

Add -h flag to surpress output files we do not need.

The 'BAM_microsats_select.sh' script subsequently selects loci which the desired repeat length. 

## Step 2. Extract selected microsatellites from genomes 

In traditional microsatellite studies, PCR serves to selectively amplify the regions of interest, namely the microsat loci.  
Downsampling bam-files by only keeping reads which overlap with the selected microsatellites.

## Step 3. Allele scoring

![alt text](https://github.com/mennodejong1986/MicrosatelliteGenotypingFromBam/blob/main/Microsatellite_genotyping_step3.png)
***Figure 1. Microsatellite scoring.*** *Reads with microsatellites*

## Step 4. Genotype calling

Distinguishing true signals from stutter alleles and null alleles.

![alt text](https://github.com/mennodejong1986/MicrosatelliteGenotypingFromBam/blob/main/Microsatellite_genotyping_step4.png)

## Step 5. Data filtering





